components/automate-ui/src/app/pages/policy/add-members/policy-add-members.component.ts

```diff
@@ -305,4 +327,79 @@ export class PolicyAddMembersComponent implements OnInit, OnDestroy {
     this.addOrRemoveQueuedMember(true, member);
     this.closeModal();
   }
+
+  private showInputs(inputName: string): void {
+
+    const formValues = this.expressionForm.value;
+    const matchAllWildCard = '*';
+
+    this.setFormLabels(formValues.type);
+
+    switch (inputName) {
+      case 'type':
+        this.resetFormControls();
+        if (formValues.type === 'user' || formValues.type === 'team') {
+          this.addIdentityControl();
+        } else if (formValues.type === 'token') {
+          this.addNameControl();
+        }
+        break;
+      case 'identity':
+        if (formValues.identity !== matchAllWildCard) {
+          this.addNameControl();
+        } else {
+          this.expressionForm.removeControl('name');
+        }
+        break;
+      default:
+        break;
+    }
+
+  }
+
+  private setFormLabels(typeValue): void {
+    if (typeValue === 'token') {
+      this.nameOrId = 'ID';
+    } else {
+      this.allIdentity = typeValue;
+      this.nameOrId = 'Name';
+    }
+  }
+
+  private resetFormControls(): void {
+    this.expressionForm.removeControl('identity');
+    this.expressionForm.removeControl('name');
+  }
+
+  private resetForm(): void {
+    this.resetFormControls();
+    this.expressionForm.reset();
+    this.expressionOutput = '';
+  }
+
+  private addIdentityControl(): void {
+    this.expressionForm.addControl('identity', new FormControl('', Validators.required));
+  }
+
+  private addNameControl(): void {
+    this.expressionForm.addControl('name', new FormControl('',
+      [
+        Validators.required,
+        Validators.pattern(Regex.patterns.NON_BLANK),
+        Validators.pattern(Regex.patterns.NO_MIXED_WILDCARD)
+      ]
+    )
+    );
+  }
+
+  private setExpressionOutput(): void {
+    const values: string[] = Object.values(this.expressionForm.value);
+    const output = values.filter(value => value != null && value.length > 0);
+    this.expressionOutput = output.join(':');
```

Comment:
This is really personal preference so feel free to ignore this note; functional style is a good pattern to be aware of, though.
You will see this a lot in our *.effects.ts files, or example.
Here, you could write this block as:
```
    this.expressionOutput =
      (Object.values(this.expressionForm.value) as string[])
      .filter(value => value != null && value.length > 0)
      .join(':');
 ```

Reply:
thanks for showing, I think i'll update to use this style